### DISCUSSION TITLE
Type checking with extra fields in construct

### DISCUSSION THREAD

**Discussion by markedwards** [pydantic/pydantic#7219] [Pydantic V2]
Currently this does not raise any type error:

```python
class Foo(BaseModel):
    foo: str


foo = Foo.model_construct(foo="foo", bar="bar")
```

I would expect a type checking error due to the use of the non-existent `bar` field.

Is this possible to achieve?  Has this behavior changed in pydantic v2?

**Comment by RS-Credentive** [comment]
I see two potential issues here:

1. According to [the documentation on Extra Fields](https://docs.pydantic.dev/latest/usage/models/#extra-fields), "By default, Pydantic models won't error when you provide data for unrecognized fields, they will just be ignored". That section gives guidance on changing this behavior.
1. [model_construct](https://docs.pydantic.dev/latest/usage/models/#creating-models-without-validation) creates a model without any validation, so even if you configured it to restrict, using model_construct would not honor the restrictions you put in place. To get those restrictions, you should just use `foo=Foo(foo="foo", bar="bar")`. Of course, the extra value will not raise an error unless you update the `model_config` :)

**Comment by markedwards** [comment]
But `model_construct()` does throw type errors on incorrect field usage (i.e., passing str where an int is expected), and on missing fields.  It just does not call out extra fields, even if `extra='forbid'` is set.

Normal construction like `Foo()` does zero type checking, so that's not an option.  What I want is full type-checking on internal model construction.  My recollection is that `construct()` used to call out extra fields in Pydantic v1, but I could be remembering wrong.

**Reply by RS-Credentive** [reply]
This is a good point. `model_construct()` won't run any validators, but I hadn't considered whether it would do basic type checking. 

I have confirmed with my own testing that normal construction (e.g. `Foo()` does impose) all restrictions for the code I have implemented, but you can also use `Foo.model_validate(obj)` if you want to be sure.

I do think that the behavior of extra fields changed from v1 to v2, but I can't find the statement about this in the migration guid.

**Reply by harunyasar** [reply]
> But model_construct() does throw type errors on incorrect field usage (i.e., passing str where an int is expected), and on missing fields.

How come? Is it in V1? If so, I don't know. If not, I couldn't reproduce what you said.

**Reply by RS-Credentive** [reply]
True, I assumed OP had tested - I see the same as you, @harunyasar  - `model_construct` does not reject invalid types or extra fields. Extra fields are silently dropped and bad input (wrong type) are allowed.

Normal method construction will allow extra fields (silently dropped) but reject assignment of bad types - I assume it will apply all validation, but I haven't tested that:

```py
>>> from pydantic import BaseModel
>>> class Foo(BaseModel):
...  foo: str
... 
>>> foo = Foo.model_construct(foo="foo", bar="bar")
>>> foo.model_dump()
{'foo': 'foo'}
>>> foo = Foo.model_construct(foo=1, bar="bar")
>>> foo.model_dump()
/home/vscode/.local/lib/python3.11/site-packages/pydantic/main.py:301: UserWarning: Pydantic serializer warnings:
  Expected `str` but got `int` - serialized value may not be as expected
  return self.__pydantic_serializer__.to_python(
{'foo': 1}
>>> foo = Foo(foo="foo", bar="bar")
>>> foo.model_dump()
{'foo': 'foo'}
>>> foo = Foo(foo=1, bar="bar")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/vscode/.local/lib/python3.11/site-packages/pydantic/main.py", line 159, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for Foo
foo
  Input should be a valid string [type=string_type, input_value=1, input_type=int]
    For further information visit https://errors.pydantic.dev/2.2/v/string_type 
```

@markedwards do you want something that will reject wrong types but not apply any other validation? Maybe you could override `model_construct` in a subclass of BaseModel? Could you try to catch the user exception thrown from model_dump?

**Reply by markedwards** [reply]
This has nothing to do with pydantic validation.  `model_construct()` does no validation, as it should be.  This is mypy type checking I'm talking about here, in conjunction with the pydantic mypy plugin.

**Reply by harunyasar** [reply]
> You should only ever use the model_construct() method with data which has already been validated, or that you definitely trust.

The behaviour you pointed above doesn't look wrong to me because of this warning from the documentation. From my point of view, `model_dump` converts your model against your correct types into a dictionary so no validation doesn't mean that you could dump it with no validation to a dictionary but you can construct your model and use it. In short, if your data has never been validated before so there is no way to use it by `model_dump()`.

**Reply by markedwards** [reply]
I'm not trying to validate the data here.  I'm type checking the internal code to detect typos in constructing model instances.  This has nothing to do with validating data.

**Reply by harunyasar** [reply]
What I am saying is that `model_construct` does what it should do. If you would like your model to throw an error for an extra field, `model_construct` is not the correct method. But I can understand what you are expecting to see. If you model is configured with `extra="forbid"` `model_construct` should reflect the same behaviour. IMO, it seems to me that if it could, that would go against what `model_construct` is essentially supposed to do.

**Reply by markedwards** [reply]
I don't want the behavior of `model_construct()` to change at all.  Its behavior is exactly correct.

What's at issue here is the behavior of the pydantic mypy plugin.  Are you saying it should not raise a type error if a value with an incorrect type is passed to a field?  Or if a required field is omitted?  Because it does those things (and they are VERY useful).

**Reply by harunyasar** [reply]
I am simply asking how a `classmethod` that expects an arbitrary number of keyword arguments defined as the `Any` type could raise a type error through `mypy`. If I get it correct, it is not because of mypy plugin it is because of the method signature. The idea is that you should be able to pass whatever you want to pass, regardless of its type, because `model_construct` assumes that the input you are passing has already been validated, which means that you are passing correct values with correct types. If this is not the case, `model_construct` will happily accept whatever you pass. If you still think that I am getting your issue wrong I will accept that I am already lost 🤣

**Reply by markedwards** [reply]
Have you used the pydantic mypy plugin?  This is what it does.

**Reply by harunyasar** [reply]
> Have you used the pydantic mypy plugin? This is what it does.

I believe I got you now.

```
[mypy]
plugins = pydantic.mypy


[pydantic-mypy]
init_forbid_extra = True
```

error: Unexpected keyword argument "bar" for "model_construct" of "Foo"  [call-arg]

This is what you are looking for.

**Reply by markedwards** [reply]
Yes, I use that.  But either I was hallucinating yesterday, or this was not working.  In any case I can confirm it actually does work (see final comments below) 😃

**Comment by markedwards** [comment]
I actually do not want validation.  I want type checking when constructing instances internally, to catch mistakes at dev time.  For external input, I want validation, but for internal construction I want type checking.

**Reply by RS-Credentive** [reply]
Is dev time type checking not handled by your IDE? Maybe I misunderstand the question.

**Comment by markedwards** [comment]
Okay.  Let me be clear.  I'm talking about mypy type checking with the pydantic mypy plugin.  I don't expect or need any runtime changes whatsoever.  I'm talking about what _type errors_ are raised at development time, by mypy.

**Reply by markedwards** [reply]
And to reiterate, the current behavior in V2 is:
1. passing a value of an incorrect type to a field raises a type error
2. omitting a required field raises a type error
3. passing a non-existent (extra) field, DOES NOT raise a type error

3 is my concern.  I would expect this to happen.

**Comment by markedwards** [comment]
Weeeeird.  Apparently I'm dead wrong in my assertion that forbidden extra args are not picked up by the pydantic mypy plugin.  It does indeed work (as I recall it did in v1).  I must have been smoking crack, but I could swear it was not working yesterday.  🤪  Possibly I had confused VSCode temporarily or something.

Just for reference, here are some screenshots to demonstrate what I'm talking about here:

mypy complaining about both the incorrect type of the `foo` value, as well as the extra `baz` argument (and it would complain about the missing `bar` as well):
<img width="368" alt="Screenshot 2023-08-24 at 16 39 38" src="https://github.com/pydantic/pydantic/assets/454900/86c9a54c-b3e3-486b-8998-6cd2de2f7ea4">

<img width="729" alt="Screenshot 2023-08-24 at 16 39 52" src="https://github.com/pydantic/pydantic/assets/454900/c3e14bae-29e8-4084-8088-6f342c809ead">

When not using `model_construct()`, the incorrect type of the `foo` value passes (as expected), but the extra `baz` still generates an error:
<img width="570" alt="Screenshot 2023-08-24 at 16 43 02" src="https://github.com/pydantic/pydantic/assets/454900/ddb276a6-5225-4247-9330-88fd665a0982">

**Comment by markedwards** [comment]
Nothing to see here, working as expected....

**Reply by RS-Credentive** [reply]
Glad to hear it!
